multi-educator/src/Dscp/Educator/Launcher/Mode.hs

```diff
@@ -0,0 +1,107 @@
+
+-- | Module contains the definition of Educator's WorkMode and its implementations.
+
+module Dscp.Educator.Launcher.Mode
+    (
+      -- * Markers
+      EducatorNode
+
+      -- * Constraints
+    , EducatorWorkMode
+    , CombinedWorkMode
+
+      -- * Implementations
+    , EducatorContext (..)
+    , EducatorRealMode
+    , ecWitnessCtx
+    ) where
+
+import Control.Lens (makeLenses)
+import Loot.Base.HasLens (HasLens')
+
+import Dscp.DB.SQLite (SQLiteDB)
+import Dscp.Educator.Config (HasEducatorConfig, withEducatorConfig)
+import Dscp.Educator.Launcher.Marker (EducatorNode)
+import Dscp.Educator.Launcher.Resource (EducatorResources)
+import qualified Dscp.Launcher.Mode as Basic
+import Dscp.Resource.Keys (KeyResources)
+import Dscp.Resource.Network
+import Dscp.Rio (RIO)
+import Dscp.Util.HasLens
+import qualified Dscp.Witness as W
+
+---------------------------------------------------------------------
+-- WorkMode class
+---------------------------------------------------------------------
+
+-- | Set of typeclasses which define capabilities of bare Educator node.
+type MultiEducatorWorkMode ctx m =
+    ( Basic.BasicWorkMode m
+
+    , HasMultiEducatorConfig
+
+    , MonadReader ctx m
+
+    , HasLens' ctx (Map Text SQLiteDB)
+    , HasLens' ctx (Map Text (KeyResources EducatorNode))
+    , MonadThrow m
+    )
+
+-- | Set of typeclasses which define capabilities both of Educator and Witness.
+type MultiCombinedWorkMode ctx m =
+    ( MultiEducatorWorkMode ctx m
+    , W.FullWitnessWorkMode ctx m
+    )
+
+---------------------------------------------------------------------
+-- WorkMode implementation
+---------------------------------------------------------------------
+
+data MultiEducatorContext = MultiEducatorContext
+    { _mecResources  :: !MultiEducatorResources
+      -- ^ Resources, allocated from params.
+    , _mecWitnessCtx :: !W.WitnessContext
+    }
+
+makeLenses ''MultiEducatorContext
+deriveHasLensDirect ''MultiEducatorContext
+
+type MultiEducatorRealMode = RIO MultiEducatorContext
+
+---------------------------------------------------------------------
+-- HasLens
+---------------------------------------------------------------------
+
+deriveHasLens 'ecResources ''MultiEducatorContext ''MultiEducatorResources
+deriveHasLens 'ecWitnessCtx ''MultiEducatorContext ''W.WitnessResources
+deriveHasLens 'ecWitnessCtx ''MultiEducatorContext ''W.WitnessVariables
+deriveHasLens 'ecWitnessCtx ''MultiEducatorContext ''NetServResources
+
+
+----------------------------------------------------------------------------
+-- (Almost) Natural Transformation
+----------------------------------------------------------------------------
+
+-- | This function transforms normal workmode into multi-workmode using login.
+-- It returns Nothing is login was not found
+normalToMulti :: Text -> EducatorWorkMode a -> MultiEducatorWorkMode (Maybe a)
+normalToMulti login m = do
+    MultiEducatorContext{..} <- ask
```

Comment:
My rant against `RecordWildCards` (`Foo{..}`) continues.

You can use `NamedFieldPuns` instead, like that:
```
    MultiEducatorContext { _mecRescourses } <- ask
```
Yep, thats longer. Yep, the reader of the code can _infer_ where the name `_mecRescourses` comes from. But it breaks the reading flow.